fvm/src/call_manager/default.rs

```diff
@@ -194,14 +194,46 @@ where
             self.gas_tracker.push_limit(limit);
         }
 
-        let mut result = self.with_stack_frame(|s| {
-            s.send_unchecked::<K>(from, to, method, params, value, read_only)
-        });
+        if self.call_stack_depth >= self.machine.context().max_call_depth {
+            let sys_err = syscall_error!(LimitExceeded, "message execution exceeds call depth");
+            if self.machine.context().tracing {
+                self.trace(ExecutionEvent::CallError(sys_err.clone()));
+            }
+            return Err(sys_err.into());
+        }
 
-        // If we pushed a limit, pop it.
-        if gas_limit.is_some() {
-            self.gas_tracker.pop_limit()?;
+        self.state_tree_mut().begin_transaction();
+        self.events.begin_transaction();
+        self.state_access_tracker.begin_transaction();
+        self.call_stack_depth += 1;
+
+        let (revert, mut result) = match <<Self::Machine as Machine>::Limiter>::with_stack_frame(
+            self,
+            |s| s.limiter_mut(),
+            |s| s.send_unchecked::<K>(from, to, method, params, value, read_only),
+        ) {
+            Ok(v) => (!v.exit_code.is_success(), Ok(v)),
+            Err(e) => (true, Err(e)),
+        };
+
+        self.call_stack_depth -= 1;
+        // Return the _first_ error (if any). We don't expect any errors here anyways as all error
+        // cases are fatal.
+        if let Some(err) = [
+            // End all transactions
+            self.state_access_tracker.end_transaction(revert).err(),
```

Comment:
The previous logic with early returns was technically safe, but always made me pause and bite my nails whenever I looked at it. This version should be a bit more robust.

Reply:
Oh this took me a few seconds to read correctly but then I noticed the square bracket so it started to make sense. But yeah, these early returns looked to me like a bug honestly when I read it.